data-prepper-core/src/main/java/com/amazon/dataprepper/peerforwarder/PeerForwarderReceiveBuffer.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.peerforwarder;
+
+import com.amazon.dataprepper.model.CheckpointState;
+import com.amazon.dataprepper.model.buffer.Buffer;
+import com.amazon.dataprepper.model.buffer.SizeOverflowException;
+import com.amazon.dataprepper.model.record.Record;
+import com.google.common.base.Stopwatch;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static java.lang.String.format;
+
+/**
+ * Buffer created for each stateful processor which implements {@link com.amazon.dataprepper.model.peerforwarder.RequiresPeerForwarding}
+ * interface.
+ *
+ * @since 2.0
+ */
+public class PeerForwarderReceiveBuffer<T extends Record<?>> implements Buffer<T> {
+    private static final Logger LOG = LoggerFactory.getLogger(PeerForwarderReceiveBuffer.class);
+
+    private final int bufferSize;
+    private final int batchSize;
+    private final String pluginName;
+    private final Semaphore capacitySemaphore;
+    private final LinkedBlockingQueue<T> blockingQueue;
+    private int recordsInFlight = 0;
+
+    public PeerForwarderReceiveBuffer(final int bufferSize, final int batchSize, final String pluginName) {
+        this.bufferSize = bufferSize;
+        this.batchSize = batchSize;
+        this.pluginName = pluginName;
+        this.blockingQueue = new LinkedBlockingQueue<>(bufferSize);
+        this.capacitySemaphore = new Semaphore(bufferSize);
+    }
+
+    @Override
+    public void write(T record, int timeoutInMillis) throws TimeoutException {
+        try {
+            final boolean permitAcquired = capacitySemaphore.tryAcquire(timeoutInMillis, TimeUnit.MILLISECONDS);
+            if (!permitAcquired) {
+                throw new TimeoutException(format("Plugin [%s] - Buffer is full, timed out waiting for a slot",
+                        pluginName));
+            }
+            blockingQueue.offer(record);
+        } catch (InterruptedException ex) {
+            LOG.error("Plugin [{}] - Buffer is full, interrupted while waiting to write the record", pluginName, ex);
```

Comment:
Do we need to log the plugin name(This will be plugin ID once we support unique ID's for plugins which is tracked in #1025) which this buffer is related to? Or is this an implementation detail?

Reply:
I think logging the plugin name is helpful for debugging.